src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java

```diff
@@ -0,0 +1,78 @@
+
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ClickHouseTLPHavingOracle implements TestOracle {
+
+    private final ClickHouseProvider.ClickHouseGlobalState state;
+    private final Set<String> errors = new HashSet<>();
+
+    public ClickHouseTLPHavingOracle(ClickHouseProvider.ClickHouseGlobalState state) {
+        this.state = state;
+        ClickHouseErrors.addExpectedExpressionErrors(errors);
+        errors.add("no such column"); // FIXME why?
+        errors.add("ON clause references tables to its right");
+    }
+
+    @Override
+    public void check() throws SQLException {
+        ClickHouseSchema s = state.getSchema();
+        ClickHouseSchema.ClickHouseTables targetTables = s.getRandomTableNonEmptyTables();
+        List<ClickHouseExpression> groupByColumns = Randomly.nonEmptySubset(targetTables.getColumns()).stream()
+                .map(c -> new ClickHouseColumnReference(c, null)).collect(Collectors.toList());
+        List<ClickHouseSchema.ClickHouseColumn> columns = targetTables.getColumns();
+        ClickHouseExpressionGenerator gen = new ClickHouseExpressionGenerator(state).setColumns(columns);
+        ClickHouseSelect select = new ClickHouseSelect();
+        select.setFetchColumns(groupByColumns);
+        List<ClickHouseSchema.ClickHouseTable> tables = targetTables.getTables();
+        List<ClickHouseExpression.ClickHouseJoin> joinStatements = gen.getRandomJoinClauses(tables);
+        List<ClickHouseExpression> from = ClickHouseCommon.getTableRefs(tables, state.getSchema());
+        select.setJoinClauses(joinStatements);
+        select.setSelectType(ClickHouseSelect.SelectType.ALL);
+        select.setFromTables(from);
+        // TODO order by?
+        select.setGroupByClause(groupByColumns);
+        select.setHavingClause(null);
+        String originalQueryString = ClickHouseVisitor.asString(select);
+
+        List<String> resultSet = DatabaseProvider.getResultSetFirstColumnAsString(originalQueryString, errors,
+                state.getConnection(), state);
+
+        ClickHouseExpression predicate = gen.getHavingClause();
+        select.setHavingClause(predicate);
+        String firstQueryString = ClickHouseVisitor.asString(select);
+        select.setHavingClause(new ClickHouseUnaryPrefixOperation(predicate, ClickHouseUnaryPrefixOperation.ClickHouseUnaryPrefixOperator.NOT));
+        String secondQueryString = ClickHouseVisitor.asString(select);
+        select.setHavingClause(new ClickHouseUnaryPostfixOperation(predicate, ClickHouseUnaryPostfixOperation.ClickHouseUnaryPostfixOperator.IS_NULL, false));
+        String thirdQueryString = ClickHouseVisitor.asString(select);
+        String combinedString = firstQueryString + " UNION ALL " + secondQueryString + " UNION ALL " + thirdQueryString;
+        if (combinedString.contains("EXIST")) {
```

Comment:
I think this test for `EXIST` is not necessary. It was some workaround for `SQLite`, which I should remove asap.

Reply:
done